### DISCUSSION TITLE
Temporal alignment of poses and depth images

### DISCUSSION THREAD

**Discussion by kshitijgoel007** [ethz-asl/wavemap#58] [Q&A] [answered]
Hi all, I'm trying to figure out where in the codebase is it being ensured that the pose estimate and depth image are temporally aligned. I usually rely on https://wiki.ros.org/message_filters/ApproximateTime for this purpose. However, I'm curious if wavemap contains a better solution. @victorreijgwart Let me know if you have thoughts!

**Comment by victorreijgwart** [comment, marked as answer]
Hi Kshitij,

To find the pose of each depth image, wavemap does a TF lookup using the timestamp and TF frame name in the message's `header.stamp` and `header.frame_id` fields.

It's possible to manually specify a time offset using the input's [`time_offset` ROS param](https://ethz-asl.github.io/wavemap/pages/configuration/inputs.html#:~:text=values%20to%20meters.-,Seconds%3CFloatingPoint%3E%20time_offset%20%3D%200.f,-Time%20offset%20to), which can be useful if the message timestamps are offset with respect to the odometry source and you have a good estimate for this error. For depth cameras, for example, we use the time_offset Kalibr estimates as part of its extrinsic calibration routine. ROS TF will then interpolate the pose as needed based on what poses are available. When the odometry is estimated at a reasonable rate, this works well in our experience.

Note that you can also manually specify what sensor frame name should be used in the TF lookups by setting [`sensor_frame_id`](https://ethz-asl.github.io/wavemap/pages/configuration/inputs.html#:~:text=TF%20lookups%20fail.-,std%3A%3Astring%20sensor_frame_id,-The%20frame_id%20to), which then replaces the frame name specified in the message's `header.frame_id`.

The pose lookup is located [here in the code](https://github.com/ethz-asl/wavemap/blob/main/ros/wavemap_ros/src/input_handler/depth_image_input_handler.cc#L71-L74).

Let me know if that helps!

**Reply by kshitijgoel007** [reply]
This helps, thank you!

**Reply by victorreijgwart** [reply]
Glad to hear!
